# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
#normal way
def add_two(x):
    return x + 2

print (add_two(4))

#self invoking function - it immediately calls itself
#anonymous function, runs inline. Not saved, but can be saved to variable.
(lambda x: x+ 2)(4)

6


6

#### Saving to a Variable

In [5]:
add_three = lambda x: x+ 3

print(add_three)
add_three(4)

<function <lambda> at 0x000001AF19FA81F0>


7

#### Multiple Inputs

In [7]:
#multiple inputs with no variables
(lambda x,y,z: x * y / z)(2,4,6)

1.3333333333333333

#### Passing a Lambda into a Function

In [9]:
def multiply(f, num):
    return f(num)

multiply(lambda x: x*2, -100)

-200

#### Returning a Lambda from a Function

In [10]:
#regular function
def multiply_test(num):
    return num * 4

multiply_test(5)

20

In [27]:
# #function within a function
# def parent_func():
#     def first_child():
#         return f'I am child of {name}'
#     return first_child()

# a_child = parent_func()

# parent_func()
# a_child('Mike')

In [22]:
def return_lambda(b, c):
    return lambda a, d: a * b + c * d

my_func = return_lambda(5, 6)
my_func(2, 3)

28

#### If Statements within Lambdas

In [26]:
def some_condition(num):
    if num > 10:
        return num * 2
    else:
        num + 2

#lambda *input*: return when True if 'condition', else return when fals
lambda_condition = lambda num: num * 2 if num > 10 else num + 2

lambda_condition(18)

36

In [ ]:
my_list = [4, 7, 2, 5, 8 ,3, 6, 2, 6, 2, 6, 5]

# updated_list = [lambda x: x**2/2 for x in my_list]
# print(updated_list)

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [29]:
f = lambda numb: numb**3

f(10)

1000

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [34]:
def returns_hello(num):
    return num * 2

my_list = [4, 7, 2, 5, 8 ,3, 6, 2, 6, 2, 6, 5]

list(map(returns_hello, my_list))

[8, 14, 4, 10, 16, 6, 12, 4, 12, 4, 12, 10]

#### Using Lambda's with Map

In [36]:
my_list

[4, 7, 2, 5, 8, 3, 6, 2, 6, 2, 6, 5]

In [35]:
list(map(lambda x: x**2/2 ,my_list))

[8.0, 24.5, 2.0, 12.5, 32.0, 4.5, 18.0, 2.0, 18.0, 2.0, 18.0, 12.5]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [38]:
a_list = [10, 35, 66, 3, 6, 34, 22, 13, 87]
list(map(lambda y: y*2-1,a_list))

[19, 69, 131, 5, 11, 67, 43, 25, 173]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [40]:
def is_even(num):
    if num % 2 == 0:
        return True
    else:
        return False
    
list_b = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

list(filter(is_even, list_b))

[2, 4, 6, 8, 10]

#### Using Lambda's with Filter()

In [ ]:
def a_names(name):
    if name[0] == 'A':
        return True
    else:
        return False

In [42]:
names = ['Bob', 'Andrew', 'Annie', 'Max', 'Evan', 'Alex']
# name[0]=='A' returns same result-----------------v no need for True/False
a_names = list(filter(lambda name: True if name[0] == 'A' else False, names))
print(a_names)

['Andrew', 'Annie', 'Alex']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [52]:
#import statistics

my_lst = [41, 48, 92, 77, 5, 81, 82, 72, 68, 57, 70, 84, 56, 91, 38, 5, 29, 24, 83, 94]
mean = sum(my_lst) / len(my_lst)
print(int(mean))

#list(filter(lambda x: x > statistics.mean(num_list), num_list))

lst_num = list(filter(lambda num: num > 59, my_lst))
print(lst_num)

59
[92, 77, 81, 82, 72, 68, 70, 84, 91, 83, 94]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [57]:
from functools import reduce

list_1 = [1, 2, 3, 4, 5]
reduce(lambda x, y: x*y, list_1)

total = 1
for num in list_1:
    total *= num
print(total)

120


#### Using Lambda's with Reduce()

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [63]:
def another_func(a_num):
    print('another_func is running')
    return(f'The number is {a_num}')

def my_func(num):
    if num > 15:
        return num
    print(f'My number is {num}')
    num += 1
    return my_func(num)

print(f'The return value of my_func when I pass in 10 is {my_func(10)}')

My number is 10
My number is 11
My number is 12
My number is 13
My number is 14
My number is 15
The return value of my_func when I pass in 10 is 16


In [64]:
def add_nums(num):
    #set base case
    if num <= 1:
        print("add_nums(1) = 1")
        return num
    else:
        print(f'add_nums({num}) = {num} + add_nums({num - 1})')
        return num + add_nums(num -1)
    
add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


15

#### Writing a Factorial Function

In [68]:
# 5! = 5 * 4 * 3 * 2 * 1
def factorial(num):
    if num <=1:
        return num
    return num * factorial(num-1)

factorial(7)

5040

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [73]:
def subtracting(num):
    if num <=1:
        return num
    return num - subtracting(num-1)

subtracting(10)

5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [75]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start
        start += step
    
    
for i in my_range(50, 3):
    print(i)

3
5
7
9
11
13
15
17
19
21
23
25
27
29
31
33
35
37
39
41
43
45
47
49


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [81]:
def your_iter(num):
    while num > 0:
        yield num **2
        num -= 1       

for i in your_iter(5):
    print(i)

25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [5]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

place_lst = list(filter(lambda place: True if place.strip() else False, places))
print(place_lst)



['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [36]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
author.sort(key = lambda ln: ln.split(" ")[-1].lower())
print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [2]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
fahrenheit = list(map(lambda x: (9/5)*x[1] + 32, places ))
print(fahrenheit)

[89.6, 53.6, 111.2, 84.2]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [5]:
def sequence(n):
     if n in {0, 1}:
         return n
     return sequence(n - 1) + sequence(n - 2)

sequence(6)

8